### ISSUE TITLE
Documentation created by gtkdoc is not generated when compiling

### ISSUE THREAD

**Issue by inigomartinez** [mesonbuild/meson#1844]
Using the gtkdoc function from the gnome module, the documentation is not generated when compiling. As far as I know, the documentation is generated only on install {module_name}-doc targets.

Would it make any sense to generate it just after compilations completes?

I have been looking at meson behavior when creating the 'build.ninja' file, and the {module_name}-doc target is considered a RunTarget[0], which is not executed a compile time, so the documentation is not generated.

[0] https://github.com/mesonbuild/meson/blob/master/mesonbuild/modules/gnome.py#L706

**Comment by tp-m** [comment]
Please no :)

gtk-doc is awfully slow. If anyone wants the docs generated before install they can run the custom target.

**Comment by inigomartinez** [comment]
Wouldn't not generating docs be a workaround? I am wondering if the solution to this specific problem wouldn't be to improve gtk-doc to be faster.

A common user may not known the custom target for building documentation[0] and will not found the documentation after building the project.

https://bugzilla.gnome.org/show_bug.cgi?id=782843#c1

**Comment by TingPing** [comment]
> Wouldn't not generating docs be a workaround?

Obviously it isn't a trivial task to make gtk-doc much faster.

We should certainly add `$module_name-doc` targets to our docs and then I would suggest downstreams document that for builders of their software.

**Comment by nirbheek** [comment]
You can also pass `build_by_default: true` to the gtkdoc invocation to force it to be built by default.

**Comment by nirbheek** [comment]
Uh, whoops, I was thinking of `gnome.compile_resources()`. Maybe we should add something like that to `run_target()`s too specifically for `gnome.gtkdoc()`?

**Comment by ptomato** [comment]
What I'd specifically like is a way to specify that a test depends on the documentation being built. gtk-doc includes a `gtkdoc-check` checker which complains if you forget to document something, and it's handy to run that during the test suite.

**Comment by Jehan** [comment]
> gtk-doc is awfully slow. If anyone wants the docs generated before install they can run the custom target.

Yeah but now it makes `ninja install` the painfully slow one, since it seems  that docs are rebuilt at each and every install (even if nothing changed in the repo). On our project (GIMP), a meson build is faster that the autotools one but the meson install is 2 times slower and clearly the gtk-doc generation is partially at fault (I see a few gtk-doc commands).

Now, I'm indeed not sure if slowing down the compilation instead if ideal either. Speeding-up gtk-doc would indeed be good, but maybe we should also make sure it simply isn't called unecessarily, i.e. when the source files have not been touched, docs should simply not be regenerated (same as any other generated file).

**Comment by TingPing** [comment]
> Speeding-up gtk-doc would indeed be good, but maybe we should also make sure it simply isn't called unecessarily, i.e. when the source files have not been touched, docs should simply not be regenerated (same as any other generated file).

That would require gtk-doc changes to expose what files it depends on.

**Comment by eli-schwartz** [comment]
This also causes #5793 -- targets should never be built by install without going through ninja somehow.
